keps/sig-autoscaling/2556-circumvent-50-tag-asg-limit/README.md

```diff
@@ -0,0 +1,139 @@
+      "launchTemplate": { 
+         "id": "*string*",
+         "name": "*string*",
+         "version": "*string*"
+      },
+      "modifiedAt": *number*,
+      "nodegroupArn": "*string*",
+      "nodegroupName": "*string*",
+      "nodeRole": "*string*",
+      "releaseVersion": "*string*",
+      "remoteAccess": { 
+         "ec2SshKey": "*string*",
+         "sourceSecurityGroups": [ "*string*" ]
+      },
+      "resources": { 
+         "autoScalingGroups": [ 
+            { 
+               "name": "*string*"
+            }
+         ],
+         "remoteAccessSecurityGroup": "*string*"
+      },
+      "scalingConfig": { 
+         "desiredSize": *number*,
+         "maxSize": *number*,
+         "minSize": *number*
+      },
+      "status": "*string*",
+      "subnets": [ "*string*" ],
+      "tags": { 
+         "*string*" : "*string*" 
+      },
+      "version": "*string*"
+   }
+}
+```
+
+[Current EKS interface]: https://github.com/aws/aws-sdk-go/blob/82e096143fdfb8f52fbeb4ef78d400eea6381ccd/service/eks/eksiface/interface.go#L96
+[DescribeNodegroup API]: https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeNodegroup.html
+[here]: https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeNodegroup.html#API_DescribeNodegroup_ResponseSyntax
+
+## Implementation Proposal
+
+1. In the autoscaling loop, Cluster Autoscaler finds no nodes, so it will build the nodeTemplate from other information.
+1. Cluster Autoscaler will first get some common allocatable resource values. ([allocatable resources code][])
+1. Cluster Autoscaler will then check the ASG tags for the EKS ManagedNodegroups tags. ([ASG tags code][])
+1. If Cluster Autoscaler finds the EKS ManagedNodegroups tags, it will call methods on the ManagedNodegroupCache to get labels, taints, and some other values.
+   1. If the cache has data for the given ManagedNodegroup, it will return that data.
+   1. If the data’s TTL has expired, the DescribeNodegroup API will be called and the data will be added to the cache and returned.
+   1. If there is no data cached for the ManagedNodegroup, the DescribeNodegroup API will be called and the data will be added to the cache and returned.
+1. Then Cluster Autoscaler will pull information from the ASG tags like it already does. If duplicate keys are found, the values from the ASG tags are kept.
+1. Cluster Autoscaler will then continue as usual.
+
+[allocatable resources code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L366
+[ASG tags code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L370
+
+## Testing
+
+- [TC1] Test creating a DescribeNodegroup request
+- [TC2] Test pulling values from cache when data exists for the ManagedNodegroups
+- [TC3] Test pulling values from cache and calling DescribeNodegroup API if no data exists
```

Comment:
Should there be a test case for if the cache TTL has expired?

Reply:
I merged the cache test description in TC3. I'll separate them since there would be 2 paths to check.